mantisshrimp/models/mantis_rcnn/mantis_faster_rcnn.py

```diff
@@ -4,16 +4,89 @@
+    def __init__(
+        self,
+        n_class,
+        h=256,
+        backbone=None,
+        pretrained=True,
+        fpn=True,
+        metrics=None,
+        **kwargs,
+    ):
         super().__init__(metrics=metrics)
-        self.n_class, self.h, self.pretrained = n_class, h, pretrained
-        self.m = fasterrcnn_resnet50_fpn(pretrained=self.pretrained, **kwargs)
-        in_features = self.m.roi_heads.box_predictor.cls_score.in_features
-        self.m.roi_heads.box_predictor = FastRCNNPredictor(in_features, self.n_class)
+        self.n_class, self.h, self.backbone, self.pretrained, self.fpn = (
+            n_class,
+            h,
+            backbone,
+            pretrained,
+            fpn,
+        )
+
+        supported_resnet_models = [
+            "resnet18",
+            "resnet34",
+            "resnet50",
+            "resnet101",
+            "resnet152",
+            "resnext50_32x4d",
+            "resnext101_32x8d",
+            "wide_resnet50_2",
+            "wide_resnet101_2",
+        ]
+
+        if self.backbone is None:
+            # Creates the default fasterrcnn as given in pytorch. Trained on COCO dataset
+            self.m = fasterrcnn_resnet50_fpn(
+                pretrained=self.pretrained,
+                num_classes=self.n_classes,
+                pretrained_backbone=True,
+                **kwargs,
+            )
+            in_features = self.m.roi_heads.box_predictor.cls_score.in_features
+            self.m.roi_heads.box_predictor = FastRCNNPredictor(
+                in_features, self.n_class
+            )
+
+        elif isinstance(self.backbone, str):
+            # Giving string as a backbone, which is either supported resnet or backbone
+            if self.fpn is True:
+                # Creates a torchvision resnet model with fpn added
+                # Will need to add support for other models with fpn as well
+                # Passing pretrained True will initiate backbone which was trained on ImageNet
+                if self.backbone in supported_resnet_models:
+                    self.m = resnet_fpn_backbone(
+                        backbone_name=self.backbone, pretrained=self.pretrained
+                    )
+                    self.m = FasterRCNN(self.backbone, self.n_class, **kwargs)
+                else:
+                    raise NotImplementedError(
+                        "FPN for non resnets is not supported yet"
+                    )
+
+            else:
+                # This does not create fpn backbone, it is supported for all models
+                self.base_model = create_torchvision_backbone(
+                    backbone=backbone, pretrained=self.pretrained
+                )
+                self.m = FasterRCNN(
+                    backbone=self.base_model, num_classes=self.n_class, **kwargs
+                )
+
+        elif isinstance(self.backbone, torch.nn.modules.container.Sequential):
```

Comment:
Just a curious question, why check for `Sequential` instead of `nn.Module`?

Reply:
Nothing specific as such, Since other torchvision backbones were of this class,
```
isinstance(ft_backbone, torch.nn.Module)
True
```
Replacing it to this. It works either ways.